### PR TITLE
Add windows cmake DLL test

### PR DESCRIPTION
**Pull request by veblush**
Related to #24844

### CODE REVIEW

**Comment by veblush** [comment]
[grpc_build_artifacts_multiplatform](https://g3c.corp.google.com/results/invocations/cc48c5dc-68e8-456e-97eb-a0b0416a4108): running

**Comment by veblush** [comment]
It turns out that windows distribution tests didn't pass. ([log](https://source.cloud.google.com/results/invocations/634ca7a9-668d-48ba-a18c-b749ade69d45/targets/github%2Fgrpc/tests)) 

```
"T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\install.vcxproj" (default target) (1) ->
"T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\ALL_BUILD.vcxproj" (default target) (3) ->
"T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\check_epollexclusive.vcxproj" (default target) (5) ->
"T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\grpc.vcxproj" (default target) (7) ->
(Link target) ->
  LINK : fatal error LNK1181: cannot open input file 'Release\gpr.lib' [T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\grpc.vcxproj]


"T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\install.vcxproj" (default target) (1) ->
"T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\ALL_BUILD.vcxproj" (default target) (3) ->
"T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\grpc++_error_details.vcxproj" (default target) (14) ->
"T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\grpc_cpp_plugin.vcxproj" (default target) (15) ->
  LINK : fatal error LNK1181: cannot open input file 'Release\grpc_plugin_support.lib' [T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\grpc_cpp_plugin.vcxproj]


"T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\install.vcxproj" (default target) (1) ->
"T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\ALL_BUILD.vcxproj" (default target) (3) ->
"T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\grpc++_unsecure.vcxproj" (default target) (18) ->
"T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\grpc_unsecure.vcxproj" (default target) (19) ->
  LINK : fatal error LNK1181: cannot open input file 'Release\gpr.lib' [T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\grpc_unsecure.vcxproj]


"T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\install.vcxproj" (default target) (1) ->
"T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\ALL_BUILD.vcxproj" (default target) (3) ->
"T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\grpc_csharp_plugin.vcxproj" (default target) (21) ->
  LINK : fatal error LNK1181: cannot open input file 'Release\grpc_plugin_support.lib' [T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\grpc_csharp_plugin.vcxproj]


"T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\install.vcxproj" (default target) (1) ->
"T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\ALL_BUILD.vcxproj" (default target) (3) ->
"T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\grpc_node_plugin.vcxproj" (default target) (22) ->
  LINK : fatal error LNK1181: cannot open input file 'Release\grpc_plugin_support.lib' [T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\grpc_node_plugin.vcxproj]


"T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\install.vcxproj" (default target) (1) ->
"T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\ALL_BUILD.vcxproj" (default target) (3) ->
"T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\grpc_objective_c_plugin.vcxproj" (default target) (23) ->
  LINK : fatal error LNK1181: cannot open input file 'Release\grpc_plugin_support.lib' [T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\grpc_objective_c_plugin.vcxproj]


"T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\install.vcxproj" (default target) (1) ->
"T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\ALL_BUILD.vcxproj" (default target) (3) ->
"T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\grpc_php_plugin.vcxproj" (default target) (24) ->
  LINK : fatal error LNK1181: cannot open input file 'Release\grpc_plugin_support.lib' [T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\grpc_php_plugin.vcxproj]


"T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\install.vcxproj" (default target) (1) ->
"T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\ALL_BUILD.vcxproj" (default target) (3) ->
"T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\grpc_python_plugin.vcxproj" (default target) (25) ->
  LINK : fatal error LNK1181: cannot open input file 'Release\grpc_plugin_support.lib' [T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\grpc_python_plugin.vcxproj]


"T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\install.vcxproj" (default target) (1) ->
"T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\ALL_BUILD.vcxproj" (default target) (3) ->
"T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\grpc_ruby_plugin.vcxproj" (default target) (26) ->
  LINK : fatal error LNK1181: cannot open input file 'Release\grpc_plugin_support.lib' [T:\src\github\grpc\workspace_cpp_windows_x86_cmake_dll\cmake\build\grpc_ruby_plugin.vcxproj]

    768 Warning(s)
    9 Error(s)
```

All errors above is about the missing `grpc_plugin_support` which is defined as

```
    {
      'target_name': 'grpc_plugin_support',
      'type': 'static_library',
      'dependencies': [
      ],
      'sources': [
        'src/compiler/cpp_generator.cc',
        'src/compiler/csharp_generator.cc',
        'src/compiler/node_generator.cc',
        'src/compiler/objective_c_generator.cc',
        'src/compiler/php_generator.cc',
        'src/compiler/python_generator.cc',
        'src/compiler/ruby_generator.cc',
      ],
    },
```

@jtattermusch Do you have an idea why this fails?

**Comment by jtattermusch** [comment]
@nicolasnoble  might have some ideas why providing a DLL build of gRPC might be problematic.

**Comment by veblush** [comment]
@jtattermusch Actually it appears that we provide the way for it? [doc](https://github.com/grpc/grpc/blob/master/BUILDING.md#windows-using-visual-studio-2015-or-2017)

**Comment by nicolasnoble** [comment]
The concept of Windows DLLs exporting containing C++ is hopelessly and fatally broken due to two main reasons:

1. There is no official C++ ABI.
This means that if you compile your DLL using Visual Studio's compiler, it's absolutely impossible to use it with a binary compiled using msys2's gcc for instance.

Say you're creating a dll that contains the following code:
```c++
demo.h:
struct Foo { int bar(); };

demo.cc:
#include "demo.h"
int Foo::bar() { return 42; }
``` 

Using Visual Studio, this will generate a dll that exports the mangled symbol "?bar@Foo@@QEAAHXZ" while under gcc, this will generate an import against the mangled symbol "_ZN3Foo3barEv".

And this doesn't even begin covering the compiler's differences in things like virtual tables or exceptions.

2. There is a stricter memory ownership issue
You can't safely allocate memory in one DLL, and free it in another. While this is somewhat doable to guarantee in C, there's absolutely no proper way to do this properly in C++, and our C++ API doesn't even begin trying to do such thing.

Both of these issues can create an influx of support requests from users who are trying to do things that their compiler and platform shouldn't allow them to do.

A good chunk of Oracle's OCCI prebuilt library forum is occupied by support requests just because of these two issues above.

**Comment by nicolasnoble** [comment]
And yes, it *is* doable to make all of this work, under certain conditions and constrains. It doesn't mean that we *should*.

**Comment by jtattermusch** [comment]
It looks like supporting shared library builds on Windows is tricky at least. Instead of adding a cmake test for building shared libraries, should we rather update our documentation to say that building DLLs is only support at "best-effort" level and that there's inherently problems with DLLs on windows and recommend that everyone links gRPC on windows statically?
I don't think we have the capacity to properly support shared library builds on windows with all their quirks.

**Comment by IIFE** [comment]
Hi all,

I've come across several issues on the gRPC repo with regards to building gRPC (C++ layer specifically) as a DLL on Windows. I've read the responses from the maintainers on the reasons gRPC C++ is not being built as a DLL. While the pointed out issues can happen, I don't think they are show stoppers for building gRPC as a DLL.

Yes C++ does not have a stable ABI, and passing memory between DLLs can cause issues. However, this is only applicable if you you're building DLLs using different compiler versions (e.g. building one DLL with Visual Studio 2013, and another with 2015). If all your code/libraries (including gRPC) are built using the same compiler/toolset, then those issues will not occur.

With tools like vcpkg it's even easier to build external libraries against the same compiler that someone uses for their own code. Additionally, from Visual Studio 2015 Microsoft have made the Visual Studio compiler backward compatible (see [here](https://docs.microsoft.com/en-us/cpp/porting/binary-compat-2015-2017?view=msvc-160#:~:text=The%20ABIs%2C%20object%20formats%2C%20and,14%20for%20all%20three%20versions.)).

In summary, I think it's very much worthwhile allowing gRPC C++ to be built as a DLL, and let the user decide if they want that option or not. You can certainly document the issues you pointed out as a precaution for the users, but they are most certainly not show stoppers and there are ways to overcome them.

Not all applications compile their codebase into a single exe, several apps use DLLs. Trying to compile gRPC static libs into several DLLs is not feasible, and people end up having to create workarounds (e.g. linking gRPC into a single DLL and creating a PIMPL for the gRPC classes).

Please consider allowing gRPC C++ to be built as a DLL, it will make life easier.

**Comment by jtattermusch** [comment]
> Hi all,
> 
> I've come across several issues on the gRPC repo with regards to building gRPC (C++ layer specifically) as a DLL on Windows. I've read the responses from the maintainers on the reasons gRPC C++ is not being built as a DLL. While the pointed out issues can happen, I don't think they are show stoppers for building gRPC as a DLL.
> 
> Yes C++ does not have a stable ABI, and passing memory between DLLs can cause issues. However, this is only applicable if you you're building DLLs using different compiler versions (e.g. building one DLL with Visual Studio 2013, and another with 2015). If all your code/libraries (including gRPC) are built using the same compiler/toolset, then those issues will not occur.
> 
> With tools like vcpkg it's even easier to build external libraries against the same compiler that someone uses for their own code. Additionally, from Visual Studio 2015 Microsoft have made the Visual Studio compiler backward compatible (see [here](https://docs.microsoft.com/en-us/cpp/porting/binary-compat-2015-2017?view=msvc-160#:~:text=The%20ABIs%2C%20object%20formats%2C%20and,14%20for%20all%20three%20versions.)).
> 
> In summary, I think it's very much worthwhile allowing gRPC C++ to be built as a DLL, and let the user decide if they want that option or not. You can certainly document the issues you pointed out as a precaution for the users, but they are most certainly not show stoppers and there are ways to overcome them.
> 
> Not all applications compile their codebase into a single exe, several apps use DLLs. Trying to compile gRPC static libs into several DLLs is not feasible, and people end up having to create workarounds (e.g. linking gRPC into a single DLL and creating a PIMPL for the gRPC classes).
> 
> Please consider allowing gRPC C++ to be built as a DLL, it will make life easier.

Thanks for sharing your thoughts. 

Note that we are not currently actively preventing gRPC C++ to be built as DLL (as you can see you can set the BUILD_SHARED_LIBS=ON in cmake), but we also don't want to actively invest into providing full support. So I think currently, the DLL build for windows is provided at a "best effort" basis and it may be occasionally broken
(at which point community would need to come back with fixes). 

@veblush 
We should make the support level for DLLs clear in our documentation:
- DLL build is supported at a "best effort" basis and we don't recommend using gRPC C++ as a DLL
- there are some known drawbacks with using windows C++ DLLs (see https://github.com/grpc/grpc/pull/25106#issuecomment-756918695) and users that decide to use them should be aware of the problems and limitations ("you've been warned")
- users are free to use it an we accept patches, but we don't guarantee it always work
- we don't have extensive testing for DLL builds in place (maintenance costs, increased test duration etc.) so regressions might occur.

**Comment by jtattermusch** [comment]
Updating the documentation on DLL builds here: https://github.com/grpc/grpc/pull/25300

**Comment by jtattermusch** [comment]
I think we should merge https://github.com/grpc/grpc/pull/25300 and  perhaps retire this PR.

**Comment by IIFE** [comment]
Thanks @jtattermusch 

Is the `BUILD_SHARED_LIBS` flag in a working state? As in does the repo build without errors when that flag is set? I'm not referring to the gotchas for using DLLs with C++/Windows, but rather if gRPC actually builds with that flag set?

I'm using `vcpkg` to build gRPC, and that flag is explicitly ignored by `vcpkg`, I think because of the suggestions from gRPC not to build it as DLLs.

**Comment by jtattermusch** [comment]
> Thanks @jtattermusch
> 
> Is the `BUILD_SHARED_LIBS` flag in a working state? As in does the repo build without errors when that flag is set? I'm not referring to the gotchas for using DLLs with C++/Windows, but rather if gRPC actually builds with that flag set?

I'd expect that a basic build would pass (but not sure what exactly you need so it's hard to tell the outcome in your case). We don't run continuous test for it so I don't know that for sure. Why don't you try it yourself?
> 
> I'm using `vcpkg` to build gRPC, and that flag is explicitly ignored by `vcpkg`, I think because of the suggestions from gRPC not to build it as DLLs.

Ok
